src/qibojit/backends/cpu.py

```diff
@@ -236,12 +237,19 @@ def _apply_ygate_density_matrix(self, gate, state, nqubits):
 
     def apply_channel_density_matrix(self, channel, state, nqubits):
         state = self.cast(state)
+        if isinstance(channel, ReadoutErrorChannel) is True:
+            state_copy = self.cast(state, copy=True)
         new_state = (1 - channel.coefficient_sum) * state
         for coeff, gate in zip(channel.coefficients, channel.gates):
             state = self.apply_gate_density_matrix(gate, state, nqubits)
             new_state += coeff * state
             # reset the state
-            state = self.apply_gate_density_matrix(gate, state, nqubits, inverse=True)
+            if isinstance(channel, ReadoutErrorChannel) is True:
```

Comment:
That's fine, but
```suggestion
            if isinstance(channel, ReadoutErrorChannel):
```
should also work.

Reply:
It should, but `pylint` throws a warning if the `is True` isn't there, that's why I included it. However, I don't know how much we care about `pylist` warnings...

Reply:
> Code looks good to me, but unfortunately I cannot confirm that tests pass on GPU due to #113. If someone can run the tests here and in qibo including GPU (and ideally both cupy and cuquantum), this should be okay to merge.

I can confirm that now the tests are working for `numba` and `cupy`, though I haven't tried with `cuQuantum`